### ISSUE TITLE
Using structs in oneOf field in Protocol Buffers v3 via Go

### ISSUE THREAD

**Issue by delaneyj** [golang/protobuf#330]
The gist is using structs in a oneOf field appears to not encode anything.

I made a [StackOverflow Question](https://stackoverflow.com/questions/43169640/using-structs-in-oneof-field-in-protocol-buffers-v3-via-go) because I'm not sure if its a bug or me not understanding the .proto v3 language spec.  New to GO and protobuf in general.

**Comment by ilifei** [comment]
I want to know how to solve it.

**Comment by cybrcodr** [comment]
I read the post.  I'm not sure if the question is (1) how to have an array of EventWrapper or (2) how to access oneOf fields?

For (1), you'll need to define a message containing repeated EventWrapper.

```
message Foo {
    repeated EventWrapper events = 1;
}
```

Note that you cannot have repeated directly on the oneof field.  The wrapper is the correct way to go.

For (2), the godoc shows example of accessing oneof fields ...
https://godoc.org/github.com/golang/protobuf/proto

```
var ew *pb_test.EventWrapper
// unmarshal JSON or other serialized form into event here.
switch e := ew.Event.(type) {
case *pb_test.EventWrapper_SignOff:
    // e.SignOff contains the SignOff object.
case *pb_test.EventWrapper_SignOn:
    // e.SignOn contains the SignOn object.
}
```

You can also use ew.GetSignOn() or ew.GetSignOff() to get the object that is set.  It will be nil if not set.

**Comment by delaneyj** [comment]
@cybrcodr isn't wasn't so much a question that a verification of errors.  Using this package wasn't able to generate repeated structs.  With [gogo](https://github.com/gogo/protobuf) had no issues.  The reference implementation not working seemed odd.

**Comment by cybrcodr** [comment]
Sorry, I think I'm missing some context here.  Looking at the proto definition you posted in StackOverflow as shown here ...
```
syntax = "proto3";

package test;

import "google/protobuf/timestamp.proto";

message Metadata {
    uint64 userID = 2;
    google.protobuf.Timestamp time= 3; 
}

//SignOff when user logs out of Glory
message SignOff {
    Metadata metadata =1;
}

//SignOn when user logs into Glory
message SignOn {
    Metadata metadata =1;
}

message EventWrapper {
    oneof event {
        SignOff signOff = 1;
        SignOn signOn = 2;
    }
}
```

I don't see any repeated fields there, and am puzzled how gogo is able to produce repeated structs.  And as mentioned, if you need to have the repeated oneof, you'll need to wrap that in another message.

Now, looking back at your StackOverflow post, you mentioned "The output shows the JSON is right but the protobuf encoded byte array is empty.".  I'm guessing this is probably what you are asking about.  I'm unable to reproduce this case though, using the same code that you posted, the return value from proto.Marshal shows a slice of length 21 for me.

**Comment by delaneyj** [comment]
Sorry, its been a few weeks.  IIRC the core issue is golang/protobuf wasn't able to generate repeated oneof fields.

**Comment by cybrcodr** [comment]
Not a problem.  I'll mark this issue as closed as I'm unable to reproduce the issue.  Feel free to reopen if you encounter the same issue again or open another one if something else comes up.  Thanks.

**Comment by delaneyj** [comment]
@cybrcodr was revamping some code and didn't use the gogo implementation and rediscovered the issue.  Looks like the root cause is when you `Marshal` a repeated field of oneOf its not actually serializing the contents at all.  Confirmed by using json & gogo implementation.

**Comment by cmoad** [comment]
I encountered this issue today. I'm curious why this issue is closed? Is it documented that you cannot serialize a repeated wrapper of oneof?

**Comment by justinsb** [comment]
I was also encountering this issue, which took me a while to figure out.  Turned out to be entirely PEBKAC, but just in case others are hitting the same thing:

I was importing `github.com/gogo/protobuf/proto` (accidentally), but I had generated my proto files against `github.com/golang/protobuf/proto`.  `proto.EncodeMessage` and `proto.Marshal` were returning empty, but there were no compile time or run-time errors.
